File: iceoryx2-cxx/tests/src/service_pipeline_tests.cpp
// Copyright (c) 2026 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache Software License 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
// which is available at https://opensource.org/licenses/MIT.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

#include "iox2/attribute_specifier.hpp"
#include "iox2/node.hpp"
#include "iox2/service.hpp"
#include "iox2/service_builder_pipeline_error.hpp"
#include "iox2/type_variant.hpp"

#include "test.hpp"

namespace {
using namespace iox2;

struct HeaderA {
    uint64_t value { 0 };
};

struct HeaderB {
    uint32_t value { 0 };
};

template <typename T>
class ServicePipelineTest : public ::testing::Test {
  public:
    static constexpr ServiceType TYPE = T::TYPE;
};

TYPED_TEST_SUITE(ServicePipelineTest, iox2_testing::ServiceTypes, );

TYPED_TEST(ServicePipelineTest, created_service_does_exist) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    const auto service_name = iox2_testing::generate_service_name();
    ASSERT_FALSE(Service<SERVICE_TYPE>::does_exist(service_name, Config::global_config(), MessagingPattern::Pipeline)
                     .value());

    auto node = NodeBuilder().create<SERVICE_TYPE>().value();

    {
        auto sut = node.service_builder(service_name).template pipeline<uint64_t>().create().value();
        ASSERT_TRUE(
            Service<SERVICE_TYPE>::does_exist(service_name, Config::global_config(), MessagingPattern::Pipeline)
                .value());
    }

    ASSERT_FALSE(Service<SERVICE_TYPE>::does_exist(service_name, Config::global_config(), MessagingPattern::Pipeline)
                     .value());
}

TYPED_TEST(ServicePipelineTest, service_name_works) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    const auto service_name = iox2_testing::generate_service_name();
    auto node = NodeBuilder().create<SERVICE_TYPE>().value();
    auto sut = node.service_builder(service_name).template pipeline<uint64_t>().create().value();

    ASSERT_THAT(sut.name().to_string().unchecked_access().c_str(),
                StrEq(service_name.to_string().unchecked_access().c_str()));
}

TYPED_TEST(ServicePipelineTest, creating_existing_service_fails) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    const auto service_name = iox2_testing::generate_service_name();
    auto node = NodeBuilder().create<SERVICE_TYPE>().value();
    auto sut = node.service_builder(service_name).template pipeline<uint64_t>().create().value();
    auto sut_2 = node.service_builder(service_name).template pipeline<uint64_t>().create();

    ASSERT_FALSE(sut_2.has_value());
    ASSERT_THAT(sut_2.error(), Eq(PipelineCreateError::AlreadyExists));
}

TYPED_TEST(ServicePipelineTest, open_or_create_service_does_exist) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    const auto service_name = iox2_testing::generate_service_name();
    auto node = NodeBuilder().create<SERVICE_TYPE>().value();

    ASSERT_FALSE(Service<SERVICE_TYPE>::does_exist(service_name, Config::global_config(), MessagingPattern::Pipeline)
                     .value());
    {
        auto sut =
            bb::Optional<PortFactoryPipeline<SERVICE_TYPE, uint64_t, void>>(node.service_builder(service_name)
                                                                          .template pipeline<uint64_t>()
                                                                          .open_or_create()
                                                                          .value());
        ASSERT_TRUE(
            Service<SERVICE_TYPE>::does_exist(service_name, Config::global_config(), MessagingPattern::Pipeline)
                .value());

        auto sut_2 =
            bb::Optional<PortFactoryPipeline<SERVICE_TYPE, uint64_t, void>>(node.service_builder(service_name)
                                                                          .template pipeline<uint64_t>()
                                                                          .open_or_create()
                                                                          .value());
        ASSERT_TRUE(
            Service<SERVICE_TYPE>::does_exist(service_name, Config::global_config(), MessagingPattern::Pipeline)
                .value());

        sut.reset();
        ASSERT_TRUE(
            Service<SERVICE_TYPE>::does_exist(service_name, Config::global_config(), MessagingPattern::Pipeline)
                .value());
        sut_2.reset();
    }

    ASSERT_FALSE(Service<SERVICE_TYPE>::does_exist(service_name, Config::global_config(), MessagingPattern::Pipeline)
                     .value());
}

TYPED_TEST(ServicePipelineTest, opening_non_existing_service_fails) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    const auto service_name = iox2_testing::generate_service_name();
    auto node = NodeBuilder().create<SERVICE_TYPE>().value();

    auto sut = node.service_builder(service_name).template pipeline<uint64_t>().open();
    ASSERT_FALSE(sut.has_value());
    ASSERT_THAT(sut.error(), Eq(PipelineOpenError::DoesNotExist));
}

TYPED_TEST(ServicePipelineTest, opening_existing_service_works) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    const auto service_name = iox2_testing::generate_service_name();
    auto node = NodeBuilder().create<SERVICE_TYPE>().value();
    auto sut_create = node.service_builder(service_name).template pipeline<uint64_t>().create().value();
    auto sut = node.service_builder(service_name).template pipeline<uint64_t>().open();

    ASSERT_TRUE(sut.has_value());
}

TYPED_TEST(ServicePipelineTest, opening_existing_service_with_wrong_payload_type_fails) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    const auto service_name = iox2_testing::generate_service_name();
    auto node = NodeBuilder().create<SERVICE_TYPE>().value();
    auto sut_create = node.service_builder(service_name).template pipeline<uint64_t>().create().value();
    auto sut = node.service_builder(service_name).template pipeline<double>().open();

    ASSERT_FALSE(sut.has_value());
    ASSERT_THAT(sut.error(), Eq(PipelineOpenError::IncompatiblePayloadType));
}

TYPED_TEST(ServicePipelineTest, opening_existing_service_with_wrong_user_header_type_fails) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    const auto service_name = iox2_testing::generate_service_name();
    auto node = NodeBuilder().create<SERVICE_TYPE>().value();
    auto sut_create = node.service_builder(service_name)
                          .template pipeline<uint64_t>()
                          .template user_header<HeaderA>()
                          .create()
                          .value();
    auto sut = node.service_builder(service_name)
                   .template pipeline<uint64_t>()
                   .template user_header<HeaderB>()
                   .open();

    ASSERT_FALSE(sut.has_value());
    ASSERT_THAT(sut.error(), Eq(PipelineOpenError::IncompatibleUserHeaderType));
}

TYPED_TEST(ServicePipelineTest, service_builder_configuration_works) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;
    constexpr uint64_t NUMBER_OF_STAGES = 3U;
    constexpr uint64_t MAX_IN_FLIGHT_SAMPLES = 23U;
    constexpr uint64_t MAX_NODES = 7U;
    constexpr uint64_t INITIAL_MAX_SLICE_LEN = 42U;

    const auto service_name = iox2_testing::generate_service_name();
    auto node = NodeBuilder().create<SERVICE_TYPE>().value();
    auto sut = node.service_builder(service_name)
                   .template pipeline<uint64_t>()
                   .number_of_stages(NUMBER_OF_STAGES)
                   .max_in_flight_samples(MAX_IN_FLIGHT_SAMPLES)
                   .max_nodes(MAX_NODES)
                   .initial_max_slice_len(INITIAL_MAX_SLICE_LEN)
                   .create()
                   .value();

    auto static_config = sut.static_config();
    ASSERT_THAT(static_config.number_of_stages(), Eq(NUMBER_OF_STAGES));
    ASSERT_THAT(static_config.max_in_flight_samples(), Eq(MAX_IN_FLIGHT_SAMPLES));
    ASSERT_THAT(static_config.max_nodes(), Eq(MAX_NODES));
    ASSERT_THAT(static_config.initial_max_slice_len(), Eq(INITIAL_MAX_SLICE_LEN));
    ASSERT_THAT(static_config.payload_type_details().variant(), Eq(TypeVariant::FixedSize));

    ASSERT_THAT(sut.number_of_stages(), Eq(NUMBER_OF_STAGES));
    ASSERT_THAT(sut.number_of_ingress_ports(), Eq(0));
    ASSERT_THAT(sut.number_of_workers(0).has_value(), Eq(true));
    ASSERT_THAT(sut.number_of_workers(0).value(), Eq(0));
    ASSERT_THAT(sut.number_of_workers(NUMBER_OF_STAGES).has_value(), Eq(false));
    ASSERT_THAT(sut.number_of_egress_ports(), Eq(0));
}

TYPED_TEST(ServicePipelineTest, runtime_role_builders_and_dynamic_lists_work) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    const auto service_name = iox2_testing::generate_service_name();
    auto node = NodeBuilder().create<SERVICE_TYPE>().value();
    auto sut = node.service_builder(service_name)
                   .template pipeline<uint64_t>()
                   .number_of_stages(1)
                   .max_in_flight_samples(8)
                   .create()
                   .value();

    auto ingress = sut.ingress_builder().create().value();
    auto worker_subscriber = sut.worker_subscriber_builder(0).value().create().value();
    auto worker_publisher = sut.worker_publisher_builder(0).value().create().value();
    auto egress = sut.egress_builder().create().value();

    ASSERT_THAT(sut.worker_subscriber_builder(1).has_value(), Eq(false));
    ASSERT_THAT(sut.worker_publisher_builder(1).has_value(), Eq(false));

    ASSERT_THAT(sut.number_of_ingress_ports(), Eq(1));
    ASSERT_THAT(sut.number_of_workers(0).value(), Eq(1));
    ASSERT_THAT(sut.number_of_egress_ports(), Eq(1));

    uint64_t ingress_count = 0;
    sut.list_ingresses([&](auto) {
        ++ingress_count;
        return CallbackProgression::Continue;
    });
    ASSERT_THAT(ingress_count, Eq(1));

    uint64_t worker_count = 0;
    sut.list_workers(0, [&](auto) {
        ++worker_count;
        return CallbackProgression::Continue;
    });
    ASSERT_THAT(worker_count, Eq(1));

    uint64_t egress_count = 0;
    sut.list_egresses([&](auto) {
        ++egress_count;
        return CallbackProgression::Continue;
    });
    ASSERT_THAT(egress_count, Eq(1));

    ingress.update_connections().value();
    worker_publisher.update_connections().value();

    ingress.send_copy(11).value();
    auto work = worker_subscriber.receive().value();
    ASSERT_THAT(work.has_value(), Eq(true));
    ASSERT_THAT(work->payload(), Eq(11));

    worker_publisher.send_copy(22).value();
    auto final_sample = egress.receive().value();
    ASSERT_THAT(final_sample.has_value(), Eq(true));
    ASSERT_THAT(final_sample->payload(), Eq(22));
}

TYPED_TEST(ServicePipelineTest, open_or_create_with_attributes_works) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;
    const auto service_name = iox2_testing::generate_service_name();
    auto node = NodeBuilder().create<SERVICE_TYPE>().value();

    auto key = *Attribute::Key::from_utf8("pipeline");
    auto value = *Attribute::Value::from_utf8("enabled");

    auto attributes = AttributeSpecifier();
    attributes.define(key, value).value();

    auto verifier = AttributeVerifier();
    verifier.require(key, value).value();

    auto created =
        node.service_builder(service_name).template pipeline<uint64_t>().open_or_create_with_attributes(verifier);
    ASSERT_TRUE(created.has_value());

    auto opened = node.service_builder(service_name).template pipeline<uint64_t>().open();
    ASSERT_TRUE(opened.has_value());

    auto counter = 0;
    opened.value().attributes().iter_key_values(key, [&](auto& v) {
        EXPECT_THAT(v.unchecked_access().c_str(), StrEq(value.unchecked_access().c_str()));
        ++counter;
        return CallbackProgression::Continue;
    });
    EXPECT_THAT(counter, Eq(1));
}
} // namespace

File: iceoryx2/tests/service_pipeline_tests.rs
// Copyright (c) 2026 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache Software License 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
// which is available at https://opensource.org/licenses/MIT.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

use iceoryx2::prelude::*;
use iceoryx2::service::port_factory::pipeline::WorkerCreateError;
use iceoryx2::testing::*;
use iceoryx2_bb_testing::assert_that;
use iceoryx2_bb_testing::watchdog::Watchdog;

#[test]
fn fixed_payload_pipeline_flow_works() {
    let _watchdog = Watchdog::new();
    type ServiceType = ipc::Service;

    let service_name = generate_service_name();
    let config = generate_isolated_config();
    let node = NodeBuilder::new()
        .config(&config)
        .create::<ServiceType>()
        .unwrap();

    let pipeline = node
        .service_builder(&service_name)
        .pipeline::<u64>()
        .number_of_stages(2)
        .max_in_flight_samples(8)
        .create()
        .unwrap();

    let opened_pipeline = node
        .service_builder(&service_name)
        .pipeline::<u64>()
        .number_of_stages(2)
        .max_in_flight_samples(8)
        .open()
        .unwrap();

    assert_that!(pipeline.number_of_stages(), eq 2);
    assert_that!(opened_pipeline.number_of_stages(), eq 2);

    let ingress = pipeline.ingress_builder().create().unwrap();
    let worker_0 = pipeline.worker_builder(0).create().unwrap();
    let worker_1 = pipeline.worker_builder(1).create().unwrap();
    let egress = pipeline.egress_builder().create().unwrap();

    assert_that!(pipeline.number_of_ingress_ports(), eq 1);
    assert_that!(pipeline.number_of_workers(0), eq Some(1));
    assert_that!(pipeline.number_of_workers(1), eq Some(1));
    assert_that!(pipeline.number_of_workers(2), eq None);
    assert_that!(pipeline.number_of_egress_ports(), eq 1);

    ingress.send_copy(11).unwrap();

    let mut stage_0_work = None;
    for _ in 0..10_000 {
        stage_0_work = worker_0.receive().unwrap();
        if stage_0_work.is_some() {
            break;
        }
    }

    let mut stage_0_work = stage_0_work.expect("stage 0 must receive sample");
    *stage_0_work.payload_mut() += 1;
    stage_0_work.send().unwrap();

    let mut stage_1_work = None;
    for _ in 0..10_000 {
        stage_1_work = worker_1.receive().unwrap();
        if stage_1_work.is_some() {
            break;
        }
    }

    let mut stage_1_work = stage_1_work.expect("stage 1 must receive sample");
    *stage_1_work.payload_mut() *= 2;
    stage_1_work.send().unwrap();

    let mut final_sample = None;
    for _ in 0..10_000 {
        final_sample = egress.receive().unwrap();
        if final_sample.is_some() {
            break;
        }
    }

    let final_sample = final_sample.expect("egress must receive sample");
    assert_that!(*final_sample, eq 24);
}

#[test]
fn dynamic_payload_pipeline_flow_works() {
    let _watchdog = Watchdog::new();
    type ServiceType = ipc::Service;

    let service_name = generate_service_name();
    let config = generate_isolated_config();
    let node = NodeBuilder::new()
        .config(&config)
        .create::<ServiceType>()
        .unwrap();

    let pipeline = node
        .service_builder(&service_name)
        .pipeline::<[u8]>()
        .number_of_stages(1)
        .max_in_flight_samples(8)
        .initial_max_slice_len(16)
        .create()
        .unwrap();

    let ingress = pipeline
        .ingress_builder()
        .initial_max_slice_len(16)
        .create()
        .unwrap();
    let worker = pipeline
        .worker_builder(0)
        .initial_max_slice_len(16)
        .create()
        .unwrap();
    let egress = pipeline.egress_builder().create().unwrap();

    let sample = ingress.loan_slice_uninit(4).unwrap();
    let sample = sample.write_from_fn(|n| (n + 1) as u8);
    sample.send().unwrap();

    let mut work = None;
    for _ in 0..10_000 {
        work = worker.receive().unwrap();
        if work.is_some() {
            break;
        }
    }

    let mut work = work.expect("worker must receive sample");
    assert_that!(work.payload_mut(), eq [1, 2, 3, 4].as_slice());
    work.payload_mut()[0] = 99;
    work.send().unwrap();

    let mut final_sample = None;
    for _ in 0..10_000 {
        final_sample = egress.receive().unwrap();
        if final_sample.is_some() {
            break;
        }
    }

    let final_sample = final_sample.expect("egress must receive sample");
    assert_that!(final_sample.payload(), eq [99, 2, 3, 4].as_slice());
}

#[test]
fn creating_worker_with_invalid_stage_fails() {
    let _watchdog = Watchdog::new();
    type ServiceType = ipc::Service;

    let service_name = generate_service_name();
    let config = generate_isolated_config();
    let node = NodeBuilder::new()
        .config(&config)
        .create::<ServiceType>()
        .unwrap();

    let pipeline = node
        .service_builder(&service_name)
        .pipeline::<u64>()
        .number_of_stages(1)
        .create()
        .unwrap();

    let result = pipeline.worker_builder(1).create();
    assert_that!(result, is_err);
    assert_that!(result.err().unwrap(), eq WorkerCreateError::StageOutOfBounds);
}

#[test]
fn worker_can_discard_sample() {
    let _watchdog = Watchdog::new();
    type ServiceType = ipc::Service;

    let service_name = generate_service_name();
    let config = generate_isolated_config();
    let node = NodeBuilder::new()
        .config(&config)
        .create::<ServiceType>()
        .unwrap();

    let pipeline = node
        .service_builder(&service_name)
        .pipeline::<u64>()
        .number_of_stages(1)
        .max_in_flight_samples(8)
        .create()
        .unwrap();

    let ingress = pipeline.ingress_builder().create().unwrap();
    let worker = pipeline.worker_builder(0).create().unwrap();
    let egress = pipeline.egress_builder().create().unwrap();

    ingress.send_copy(7).unwrap();

    let mut work = None;
    for _ in 0..10_000 {
        work = worker.receive().unwrap();
        if work.is_some() {
            break;
        }
    }

    let work = work.expect("worker must receive sample");
    work.discard();

    for _ in 0..100 {
        let sample = egress.receive().unwrap();
        assert_that!(sample, is_none);
    }

    ingress.send_copy(21).unwrap();
    let mut work = None;
    for _ in 0..10_000 {
        work = worker.receive().unwrap();
        if work.is_some() {
            break;
        }
    }

    let mut work = work.expect("worker must receive second sample");
    *work.payload_mut() += 1;
    work.send().unwrap();

    let mut final_sample = None;
    for _ in 0..10_000 {
        final_sample = egress.receive().unwrap();
        if final_sample.is_some() {
            break;
        }
    }

    let final_sample = final_sample.expect("egress must receive second sample");
    assert_that!(*final_sample, eq 22);
}
